File: algorithms/sort/insertion/C/insertion.c
#include <stdio.h>
#include <stdlib.h>

void insertionSort(int *arr, int n) {
    int j, actual;
    for (int i = 1; i < n; i++) {
        actual = arr[i];
        for (j = i; (j > 0) && (actual < arr[j-1]); j--)
            arr[j] = arr[j-1];
        arr[j] = actual;
    }
}

int main() {
    int n; int *arr;
    scanf("%d", &n);

    arr = (int*) malloc((n + 1) * sizeof(int));
    for (int i = 0; i < n; i++) {
        int temp = 0;
        scanf("%d", &temp);
        arr[i] = temp;
    }

    insertionSort(arr, n);

    for (int i=0; i<n; i++)
        printf("%d ", arr[i]);

    return 0;
}

File: algorithms/sort/insertion/python/insertion.py
"""
>>> a = [1, 2, 3, 4, 5]
>>> insertion_sort(a)
>>> a
[1, 2, 3, 4, 5]
>>> a = [4, 2, 5, 1, 3]
>>> insertion_sort(a)
>>> a
[1, 2, 3, 4, 5]
"""


def insertion_sort(a):
    for i in range(1, len(a)):
        actual = a[i]
        j = i
        while j > 0 and actual < a[j-1]:
            a[j] = a[j-1]
            j -= 1
        a[j] = actual


if __name__ == "__main__":
    import doctest
    doctest.testmod()
